py/dml/ctree.py

```diff
@@ -4063,8 +4063,16 @@ def mkCast(site, expr, new_type):
         # these casts are permitted by C only if old and new are
         # the same type, which is useless
         return Cast(site, expr, new_type)
-    if isinstance(real, (TVoid, TStruct, TArray, TVector, TTraitList,
-                         TLayout, TFunction)):
+    if isinstance(real, TStruct):
+        if isinstance(old_type, TStruct) and old_type.label == real.label:
+            return expr
+        raise ECAST(site, expr, new_type)
+    if isinstance(real, TExternStruct):
+        if isinstance(old_type, TExternStruct) and old_type.id == real.id:
+            return expr
+        raise ECAST(site, expr, new_type)
```

Comment:
The relevant use I discovered of `mkCast` where this change is important is `ExpressionInitializer.assign_to` under the `deep_const` case. Since our internal use of `mkCast` is so prolific, I think this is preferable to an ad-hoc solution inside `ExpressionInitializer.assign_to`.

Reply:
OTOH, one could argue that internal mkCasts are supposed to know what they are doing, so it should generally be hard to inject a TExternStruct into them.

But that's currently an academic question, I'm all for keeping the check in mkCast, I would do the same given what our code base looks like.